pilot/pkg/controllers/ipallocate/ipallocate.go

```diff
@@ -0,0 +1,402 @@
+	v4allocator *ipAllocator
+	v6allocator *ipAllocator
+}
+
+type conflictDetectedEvent struct {
+	conflictingResourceIdentifier types.NamespacedName
+	conflictingAddress            netip.Addr
+}
+
+const (
+	controllerName = "IP Autoallocate"
+	// these are the ranges of the v1 logic, do we need to choose new ranges?
+	IPV4Prefix = "240.240.0.0/16"
+	IPV6Prefix = "2001:2::/48"
+)
+
+func NewIPAllocate(stop <-chan struct{}, c kubelib.Client) *IPAllocate {
+	client := kclient.New[*networkingv1alpha3.ServiceEntry](c)
+	allocator := &IPAllocate{
+		serviceEntryClient: client,
+		stopChan:           stop,
+		// MustParsePrefix is OK because these are const. If we allow user configuration we must not use this function.
+		v4allocator: newIPAllocator(netip.MustParsePrefix(IPV4Prefix)),
+		v6allocator: newIPAllocator(netip.MustParsePrefix(IPV6Prefix)),
+	}
+	allocator.queue = controllers.NewQueue(controllerName, controllers.WithGenericReconciler(allocator.reconcile), controllers.WithMaxAttempts(5))
+	client.AddEventHandler(controllers.ObjectHandler(allocator.queue.AddObject))
+	return allocator
+}
+
+func (c *IPAllocate) Run(stop <-chan struct{}) {
+	log.Debugf("starting %s controller", controllerName)
+	kubelib.WaitForCacheSync(controllerName, stop, c.serviceEntryClient.HasSynced)
+	// it is important that we always warm cache before we try to run the queue
+	// failing to warm cache first could result in allocating IP which are already in use
+	c.warmCache()
+	// logically the v1 behavior is a state of the world function which cannot be replicated in reconcile
+	// if we wish to replicate v1 assignment behavior for existing serviceentry it should go here and must
+	// include a change to respect the used IPs from the warmed cache begin reconcile
+	c.queue.Run(stop)
+	c.serviceEntryClient.ShutdownHandlers()
+}
+
+// warmCache reads all serviceentries and records any auto-assigned addresses as in use
+func (c *IPAllocate) warmCache() {
+	serviceentries := c.serviceEntryClient.List(metav1.NamespaceAll, klabels.Everything())
+	count := 0
+	for _, serviceentry := range serviceentries {
+		count++
+		owner := types.NamespacedName{
+			Name:      serviceentry.Name,
+			Namespace: serviceentry.Namespace,
+		}
+		log.Debugf("%s/%s found during warming", serviceentry.Namespace, serviceentry.Name)
+		if len(serviceentry.Spec.Addresses) > 0 {
+			for _, addr := range serviceentry.Spec.Addresses {
+				a, err := netip.ParseAddr(addr)
+				if err != nil {
+					log.Debugf("unable to parse address %s for %s/%s, received error: %s", addr, serviceentry.Namespace, serviceentry.Name, err.Error())
+					continue
+				}
+				// these are not assigned by us but could conflict with our IP ranges and cause issues
+				if !c.inOurRange(a) {
+					// don't need to worry about this one
+					continue
+				}
+				c.markUsedOrQueueConflict(a, owner)
+			}
+		}
+
+		if serviceentry.Status.String() != "" {
```

Comment:
We should  be able to check something more efficiently than serializing to a string. Probably not super important until the istio/api side merge though.

Reply:
actually since it is not a pointer, which is why I was checking it this way, it should have a zero value and it is probably safe not to check if it is empty

Reply:
yeah, the check isn't necessary as far as I can tell